pkg/schedule/coordinator.go

```diff
@@ -152,72 +149,195 @@ func (c *Coordinator) IsPendingRegion(region uint64) bool {
+				continue
+			}
+
+			// Check priority regions first.
+			c.checkPriorityRegions()
+			// Check suspect regions first.
+			c.checkSuspectRegions()
+			// Check regions in the waiting list
+			c.checkWaitingRegions()
+
+			key, regions = c.checkRegions(key)
+			if len(regions) == 0 {
+				continue
+			}
+			// Updates the label level isolation statistics.
+			c.cluster.UpdateRegionsLabelLevelStats(regions)
+			// Update metrics if a full scan is done.
+			if len(key) == 0 {
+				c.patrolRegionContext.roundUpdate()
+				c.patrolRegionContext.updateScanLimit(c.cluster)
+			}
+			failpoint.Inject("break-patrol", func() {
+				time.Sleep(100 * time.Millisecond) // ensure the regions are handled by the workers
+				failpoint.Return()
+			})
 		case <-c.ctx.Done():
 			patrolCheckRegionsGauge.Set(0)
-			c.setPatrolRegionsDuration(0)
+			c.patrolRegionContext.setPatrolRegionsDuration(0)
 			log.Info("patrol regions has been stopped")
 			return
 		}
-		if c.cluster.IsSchedulingHalted() {
-			continue
-		}
+	}
+}
 
-		// Check priority regions first.
-		c.checkPriorityRegions()
-		// Check suspect regions first.
-		c.checkSuspectRegions()
-		// Check regions in the waiting list
-		c.checkWaitingRegions()
+func (c *Coordinator) updateTickerIfNeeded(ticker *time.Ticker) {
+	// Note: we reset the ticker here to support updating configuration dynamically.
+	newInterval := c.cluster.GetCheckerConfig().GetPatrolRegionInterval()
+	if c.patrolRegionContext.interval != newInterval {
+		c.patrolRegionContext.interval = newInterval
+		ticker.Reset(newInterval)
+	}
+}
 
-		key, regions = c.checkRegions(key)
-		if len(regions) == 0 {
-			continue
-		}
-		// Updates the label level isolation statistics.
-		c.cluster.UpdateRegionsLabelLevelStats(regions)
-		if len(key) == 0 {
-			dur := time.Since(start)
-			patrolCheckRegionsGauge.Set(dur.Seconds())
-			c.setPatrolRegionsDuration(dur)
-			start = time.Now()
-		}
-		failpoint.Inject("break-patrol", func() {
-			failpoint.Break()
-		})
+func (c *Coordinator) updatePatrolWorkersIfNeeded() {
+	newWorkersCount := c.cluster.GetCheckerConfig().GetPatrolRegionWorkerCount()
+	if c.patrolRegionContext.workerCount != newWorkersCount {
+		log.Info("coordinator starts patrol regions with new workers count",
+			zap.Int("old-workers-count", c.patrolRegionContext.workerCount),
+			zap.Int("new-workers-count", newWorkersCount))
+		c.patrolRegionContext.workerCount = newWorkersCount
+		// Stop the old workers and start the new workers.
+		c.patrolRegionContext.workersCancel()
+		c.wg.Wait()
+		c.patrolRegionContext.workersCtx, c.patrolRegionContext.workersCancel = context.WithCancel(c.ctx)
+		c.patrolRegionContext.startPatrolRegionWorkers(c)
 	}
 }
 
 func (c *Coordinator) checkRegions(startKey []byte) (key []byte, regions []*core.RegionInfo) {
-	regions = c.cluster.ScanRegions(startKey, nil, patrolScanRegionLimit)
+	regions = c.cluster.ScanRegions(startKey, nil, c.patrolRegionContext.scanLimit)
```

Comment:
Will it be the bottleneck?

Reply:
`max(patrolScanRegionMinLimit, cluster.GetTotalRegionCount()/patrolRegionPartition)` will be suitable with big cluster